File: leetcode/april2024/07_valid_parenthesis.py
# CODE

class Solution:
    def checkValidString(self, s: str) -> bool:
        open = close = 0
        n = len(s) - 1
        for i in range(n + 1):
            if s[i] == '(' or s[i] == '*': open += 1
            else: open -= 1
            if s[n - i] == ')' or s[n - i] == '*': close += 1
            else: close -= 1
            if open < 0 or close < 0: return False
        return True

File: leetcode/april2024/07_valid_parenthesis.cpp
// CODE

class Solution {
    public:
    bool checkValidString(string s) {
        int n = s.size() - 1, open = 0, close = 0;
        for(int i = 0; i <= n; i++) {
            if(s[i] == '(' or s[i] == '*') open++;
            else open--;
            if(s[n - i] == ')' or s[n - i] == '*') close++;
            else close--;
            if(open < 0 or close < 0) return 0;
        }
        return 1;
    }
};